File: Data Structures & Algorithms/surrounded-regions/submission-3.java
class Solution {
    public void solve(char[][] board) {
        for (int i = 0; i < board.length; i++) {
            dfs(board, i, 0);
            dfs(board, i, board[0].length - 1);
        }
        for (int j = 0; j < board[0].length; j++) {
            dfs(board, 0, j);
            dfs(board, board.length - 1, j);
        }

        for (int i =0; i < board.length; i++) {
            for(int j = 0; j < board[0].length; j++){
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }else if (board[i][j] == '#') {
                    board[i][j] = 'O';
                }
            }
        }
    }
    private void dfs(char[][] board, int r, int c) {
        if (r < 0 || r >= board.length || c < 0 || c >= board[0].length || board[r][c] != 'O') {
            return;
        }
        board[r][c] = '#';
        dfs(board, r + 1, c);
        dfs(board, r - 1, c);
        dfs(board, r, c + 1);
        dfs(board, r, c - 1);
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-3.cs
public class Solution {
    public int MaxAreaOfIsland(int[][] grid) {
        int maxArea = 0;
        var visited = new HashSet<(int row, int col)>();
        for (int i =0; i < grid.Length; i++){
            for (int j = 0; j < grid[0].Length; j++) {
                if (!visited.Contains((i, j)) && grid[i][j] == 1) {
                    int area = Dfs(i, j, visited, grid);
                    maxArea = Math.Max(maxArea, area);
                }
            }
        }
        return maxArea;
    }
    private int Dfs(int r, int c, HashSet<(int row, int col)> visited, int[][] grid) {
        if (r < 0 || r >= grid.Length || c < 0 || c >= grid[0].Length || visited.Contains((r, c)) || grid[r][c] == 0) {
            return 0;
        }
        visited.Add((r, c));
        return 1 + Dfs(r + 1 , c, visited, grid) + Dfs(r - 1, c, visited, grid) + Dfs(r, c + 1, visited, grid) + Dfs(r, c- 1, visited, grid);
    }
}
